features/events/traps/src/main/java/org/opennms/netmgt/trapd/TrapReceiverImpl.java

```diff
@@ -64,10 +70,97 @@
+    private List<TrapNotificationHandler> m_trapNotificationHandlers = new ArrayList<TrapNotificationHandler>();
+
+    private Map<String,SnmpV3User> m_updatedSnmpV3Users = new TreeMap<String, SnmpV3User>();
+
+    private Map<String,SnmpV3User> m_SnmpV3UsersMap = new TreeMap<String, SnmpV3User>();
+
+
+    public void setTrapdConfig(TrapdConfiguration m_trapdConfig) {
+        m_updatedSnmpV3Users = addToSnmpV3Users(m_trapdConfig);
+        if (checkForTrapdConfigurationChange(m_trapdConfig)) {
+            stop();
+            LOG.info("TrapReceiver service has been stopped.");
+            m_snmpTrapPort = m_trapdConfig.getSnmpTrapPort();
+            m_snmpTrapAddress = m_trapdConfig.getSnmpTrapAddress();
+            m_snmpV3Users = new ArrayList<SnmpV3User>(m_updatedSnmpV3Users.values());
+            clearValues();
+            LOG.info("Starting the TrapReceiver service.");
+            start();
+            LOG.info("TrapReceiver service has been restarted successfully.");
+        }
+    }
+
+    private void clearValues() {
+        m_updatedSnmpV3Users = Collections.emptyMap();
+        m_SnmpV3UsersMap = Collections.emptyMap();
+    }
+
+    public static boolean compareSnmpV3UsersMap(Map<String, SnmpV3User> existingSnmpV3UserMap, Map<String, SnmpV3User> updatedSnmpV3Usermap) {
+
+        if (existingSnmpV3UserMap.isEmpty() || updatedSnmpV3Usermap.isEmpty()) {
+            return true;
+        }
+
+        for (String securityName : existingSnmpV3UserMap.keySet()) {
+            if (
+                compareSnmpV3UsersAttributes(existingSnmpV3UserMap.get(securityName).getAuthPassPhrase(), updatedSnmpV3Usermap.get(securityName).getAuthPassPhrase()) ||
+                compareSnmpV3UsersAttributes(existingSnmpV3UserMap.get(securityName).getAuthProtocol(), updatedSnmpV3Usermap.get(securityName).getAuthProtocol()) ||
+                compareSnmpV3UsersAttributes(existingSnmpV3UserMap.get(securityName).getEngineId(), updatedSnmpV3Usermap.get(securityName).getEngineId()) ||
+                compareSnmpV3UsersAttributes(existingSnmpV3UserMap.get(securityName).getPrivPassPhrase(), updatedSnmpV3Usermap.get(securityName).getPrivPassPhrase()) ||
+                compareSnmpV3UsersAttributes(existingSnmpV3UserMap.get(securityName).getPrivProtocol(), updatedSnmpV3Usermap.get(securityName).getPrivProtocol())
+            ) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    private static boolean compareSnmpV3UsersAttributes(String currentValue, String updatedValue) {
+        if (currentValue != null && updatedValue != null) {
+            if (!currentValue.equalsIgnoreCase(updatedValue)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    private static Map<String, SnmpV3User> listToMapConversion(List<SnmpV3User> m_snmpV3Users) {
+        Map<String, SnmpV3User> snmpV3UserMap = Collections.synchronizedMap(new ConcurrentHashMap<String, SnmpV3User>());
```

Comment:
You are using a synchronizedMap() here but there is no synchronized{} block. This is not correct.